### ISSUE TITLE
Nullable Reference Types: Compiler should not report CS8614 for a nullable value typed argument when the implemented interface lies within another assembly

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#27008]
**Version Used**: 05/14/18 Nullable Reference Types Preview (`csc` reports `2.8.0.62830 (e595ee27)`) with Visual Studio 15.7.1

**Demonstration Code**:

I've created [PathogenPlayground/TestCs8614](https://github.com/PathogenPlayground/TestCs8614) to demonstrate the issue.

**Steps to Reproduce**:

1. Define an interface in a class library with a method `void Test(int? nullableInt)`
2. Create a separate assembly that references the previous class library.
3. Create the same interface in the new assembly.
4. Create implementations for each interface.

**Expected Behavior**:

No warnings are reported.

**Actual Behavior**:

You receive the following warning, but only for the implementation that uses the interface from the other assembly:

```
warning CS8614: Nullability of reference types in type of parameter 'nullableInt' doesn't match implicitly implemented member 'void 
```

----

The readme on [PathogenPlayground/TestCs8614](https://github.com/PathogenPlayground/TestCs8614) has some additional details about how things are set up in the demonstration.

**Comment by PathogenDavid** [comment]
Since I think it might be related, here is a similar issue I've found concerning covariant return types:

[PathogenPlayground/Cs8613WithCovariantReturnType](https://github.com/PathogenPlayground/Cs8613WithCovariantReturnType)

Similar to this main issue, if the two interfaces are moved into the main assembly, the warning disappears.

**Comment by jcouv** [comment]
Tagging @cston 
I don't understand why the compiler complains about mismatching nullability of **reference types** given that the interface uses `int?` (a nullable **value type**).

**Comment by cston** [comment]
Thanks for reporting this issue. Fixed in #28105.

**Comment by PathogenDavid** [comment]
No problem, thanks for getting it fixed!

**Comment by froufrou121** [comment]
Since Visual Studio 16.2.0 Preview 4 I get CS8614 warnings in the following case:
class library C#8 #nullable enable

    public interface I1
    {
        string Name { get; }
    }

    public interface I2
    {
        I1? GetItem(int id);
    }

Second assembly (if everything is in the same assembly you do not get the warning) with a class implementing I2:
    public class C2 : I2
    {
        public I1? GetItem(int id) => null;
    }
The warning displayed in the output window is :
2>  C:\Users\frous\source\repos\TestCS8614\TestCS8614\Program.cs(15,20,15,27): warning CS8614: Nullability of reference types in type of parameter 'id' of 'I1? C2.GetItem(int id)' doesn't match implicitly implemented member 'I1? I2.GetItem(int id)'.

Also note that when you open the source file the green underline is there to start with but sometimes  disappears after a few moments as if the code is compiled with a different version of Roslyn from the UI.

I have attached the project that I have used to reproduce the issue. Please open and compile with Visual Studio 16.2.0 Preview 4.

[TestCS8614.zip](https://github.com/dotnet/roslyn/files/3403146/TestCS8614.zip)

**Comment by jcouv** [comment]
@cston Can you take a look (confirm the new repro and triage as needed)? Thanks

**Comment by cston** [comment]
The latest repro sounds like https://github.com/dotnet/roslyn/issues/37281.

**Comment by froufrou121** [comment]
Not too sure this is the same issue: here both the interface and the class implementing the interface have opted for nullables using #nullable enable. It happens only for parameters that are non nullable struct types.

I have added some members to my interface which looks like this now:

    public interface I2
    {
        I1? GetItem(int id);
        I1? GetItem1(int? id);
        I1? GetItem2(int i, int j);
        I1? GetItem3(string id);
        I1? GetItem4(double id);
    }

No warnings are reported for GetItem1 and GetItem3, 1 warning is reported for GetItem and GetItem4, 2 warnings are reported for GetItem2.
It seems that the compiler treats the int / double parameters are reference types.

**Comment by cston** [comment]
Thanks @froufrou121 for the detailed description. I was able to repro the issue with Visual Studio 16.2.0 Preview 4. The issue no longer repros with 16.3.0 Preview 1.
